cpp/src/arrow/util/compare.h

```diff
@@ -54,6 +54,10 @@ class EqualityComparable {
   bool operator==(const T& other) const { return cast().Equals(other); }
   bool operator!=(const T& other) const { return !(cast() == other); }
 
+  friend bool operator==(const T& a, const T& b) {
+    return a.Equals(b);
+  }
```

Comment:
Perhaps we can only use hidden friends.

```suggestion
  friend bool operator==(const T& a, const T& b) {
    return a.Equals(b);
  }
  friend bool operator!=(const T& a, const T& b) {
    return !(a.Equals(b));
  }
```

Reply:
It's ok on my PC, I'll try it